src/lib.rs

```diff
@@ -0,0 +1,198 @@
+extern crate regex;
+use regex::Regex;
+use std::error;
+use std::fmt;
+
+
+pub struct Siso {
+    route: Route,
+}
+
+impl Siso {
+    pub fn register(&mut self, path_nodes : &[PathNode], value: RouteValue) -> Result<bool, SisoError> {
+        if 0 == path_nodes.len() {
+            panic!("Need a least one node for registering a value...");
+        }
+        return self.route.register(path_nodes, value);
+    }
+
+    pub fn find(&self, path_nodes : &[PathNode]) -> Result<RouteValue, SisoError> {
+        if 0 == path_nodes.len() {
+            panic!("Need a least one node for finding a value...");
+        }
+        return self.route.find(path_nodes);
+    }
+
+    pub fn new() -> Siso {
+        Siso { route: Route::new(PathNode::PathPart(String::from("[root]")), None) }
+    }
+}
+
+#[derive(Debug)]
+struct Route {
+    node: PathNode,
+    routes: Vec<Route>,
+    value: RouteValue,
+}
+
+impl Route {
+    fn register(&mut self, path_nodes : &[PathNode], value: RouteValue) -> Result<bool, SisoError> {
+        let path_node = &path_nodes[0];
+        let mut index = Err("");
+
+        for i in 0..self.routes.len() {
+            if *path_node == self.routes[i].node {
+                index = Ok(i)
+            }
+        }
```

Comment:
There are a couple of ways to improve this loop:

* use `enumerate`: `for (i, route) in self.routes.iter().enumerate() {`
* use `find`: `self.routes.iter().enumerate().find(|(i, route)| ... )`

Reply:
That was the tricky part for me :). In fact, I first tried a lot of ways to iterate but if ended stuck with the borrow checker until I reached out that code.

That said, I'll try to implement it with with the solutions you recommend and if you agree, I'll ask you some help if I still have problems with the borrow checker.